code/game/machinery/mac_cannon.dm

```diff
@@ -0,0 +1,210 @@
+/obj/machinery/mac_barrel
+	name = "\improper MAC cannon barrel"
+	desc = "Make sure this is pointing the right way."
+	icon = 'icons/obj/stationobjs.dmi'
+	icon_state = "mac_barrel"
+
+	density = 1
+	anchored = 1
+
+	var/obj/machinery/mac_breech/breech = null
+	var/id = 0
+
+
+/obj/machinery/mac_barrel/New()
+	..()
+	find_breech()
+
+/obj/machinery/mac_barrel/proc/can_fire()
+	.= 0
+	if(breech.loader)
+		return
+	if(breech.charge_process < 100)
+		return
+	return 1
+
+/obj/machinery/mac_barrel/proc/attempt_fire(var/datum/starship/target)
+	if(!can_fire()) return
+	spawn toggle_hatch()
+	breech.charge_process = 0
+	playsound(breech,'sound/weapons/mac_fire.ogg',100,0)
+	if(breech.loaded_shell)
+		if(breech.loaded_objects.len > 1) // if there is a shell and other shit in the barrel, blow it up
+			explosion(breech,1,2,6)
+		else
+			var/obj/item/projectile/ship_projectile/mac_round/M = PoolOrNew(breech.loaded_shell.projectile,get_step(src,dir))
+			M.set_data(breech.loaded_shell.damage,1,0,target,breech.loaded_shell.armed)
+			M.setDir(src,dir)
+			M.starting = src.loc
+			M.fire()
+			fire_sound()
+		var/obj/item/weapon/twohanded/required/shell_casing/C = new breech.loaded_shell.casing
+		C.forceMove(breech)
+		qdel(breech.loaded_shell)
+		breech.loaded_objects -= breech.loaded_shell
+		breech.loaded_shell = null
+
+	else //otherwise shoot whatever is in the barrel
+		for(var/atom/movable/A in breech.loaded_objects)
+			A.forceMove(get_step(src,dir))
+			if(istype(A,/mob/living/))
+				var/mob/living/M = A
+				M.Weaken(5)
+			var/atom/throw_at = get_edge_target_turf(src, dir)
+			A.throw_at_fast(throw_at, 500, 1)
+
+			sleep(2)
+
+
+/obj/machinery/mac_barrel/proc/toggle_hatch() //just moves the functionality of the massdriver control into the mac cannon, cleaned up a bit
+	var/list/activated_doors = list()
+	for(var/obj/machinery/door/poddoor/M in machines)
+		if(M.id == id)
+			activated_doors += M
+			spawn(0) M.open()
+	sleep(50)
+
+	for(var/obj/machinery/door/poddoor/M in activated_doors)
+		spawn(0) M.close()
+
+
+/obj/machinery/mac_barrel/proc/fire_sound()
+	playsound(loc, 'sound/weapons/flashbang.ogg', 100, 1)
+	for(var/mob/living/carbon/human/H in range(5,src))
+		if(H.check_ear_prot()) continue
+		if(!istype(H.loc.loc,/area/shuttle/ftl/munitions)) continue
+
+		H.show_message("<span class='warning'>BANG</span>", 2)
```

Comment:
This is probably the part I hate the most. It feels so cheezy copying flashbang code.